src/Components/Web.JS/src/Boot.Server.Common.ts

```diff
@@ -119,7 +109,76 @@ export async function startCircuit(components: RootComponentManager<ServerCompon
   jsInitializer.invokeAfterStartedCallbacks(Blazor);
 }
 
-async function initializeConnection(options: CircuitStartOptions, logger: Logger, circuit: CircuitDescriptor): Promise<HubConnection> {
+export function startCircuit(components: RootComponentManager<ServerComponentDescriptor>): Promise<boolean> {
+  if (!started) {
+    throw new Error('Cannot start the circuit until Blazor Server has started.');
+  }
+
+  circuitActive = true;
+  startCircuitPromise ??= (async () => {
+    const appState = discoverPersistedState(document);
+    renderQueue = new RenderQueue(logger);
+    circuit = new CircuitDescriptor(components, appState || '');
+    dotNetDispatcher = new CircuitDotNetCallDispatcher(() => connection);
+    dispatcher = DotNet.attachDispatcher(dotNetDispatcher);
+
+    const initialConnection = await initializeConnection(logger, circuit);
+    const circuitStarted = await circuit.startCircuit(initialConnection);
+    if (!circuitStarted) {
+      logger.log(LogLevel.Error, 'Failed to start the circuit.');
+      return false;
+    }
+    return true;
+  })();
+
+  return startCircuitPromise;
+}
+
+export async function disposeCircuit() {
+  if (!circuitActive) {
+    return;
+  }
+
+  circuitActive = false;
+
+  await startCircuitPromise;
+
+  if (circuitActive) {
+    // A call to 'startCircuit' was made while we were waiting to dispose the circuit.
+    // Therefore, we should abort the disposal.
+    return;
+  }
```

Comment:
I appreciate that logic like this can be a bit hard to follow. Unfortunately, we have a recurring problem in Blazor's JS implementation where an assumption was previously made that certain state should be "static" / "singleton" / "module-level", when in fact it's actually better represented as being instance state with a limited lifetime.

In this example, it would be better if there was a more concrete concept of a "circuit" instance on the JavaScript side, where that circuit can be disposed and a new one can be created. This allows a separate copy of state to exist on each circuit instance, so you don't end up with tricky synchronization/ordering challenges during disposal/initialization.

Unfortunately, these kinds of challenges already existed in this file with the disconnection/reconnection logic. It would be possible for us to through and rework this file to use instance-level state (rather than module-level state) at the risk of us potentially changing something subtle in the existing initialization/reconnection logic behavior. Maybe it's worth doing this now, or maybe this is something we should look into more closely for .NET 9.

Reply:
Yeah, this is tricky. I do think the `Web.JS` code during .NET 8 has gone from "*of mixed quality*" to "*needs a very thorough rewrite if we're going to maintain and expand this long term*". It's always been a lower priority than the C# that runs server-side, and its smallness used to let us get away with idiosyncrasies, but it's no longer small and gets ever more spaghettified.

So I would say:

 * For .NET 8, let's follow your judgement about how much tidying is needed in the short term. We need correctness and to avoid too much weirdness that will make debugging difficult (especially given we may still be patching 8.0 three years from now).
 * For .NET 9, we should be prepared to eat the cost of a bigger renovation of the JS/TS code, even a clean start. I would hope that in something like 3-4 weeks we could go through the process of recreating the functionality with a proper set of Jest tests, copy-pasting *small* chunks from the old code to fit in with a better, instance-oriented architecture.
   * A major asset for us is the E2E tests which are really thorough in most aspects, so we should have good confidence about correctness. That's even more so now a lot of the WebAssembly startup flow is outside this repo, so we no longer have to worry about recreating all the semantics around loading and any errors that can occur.

@MackinnonBuck What do you think? If you think we should be tackling any of this sooner, or if you think a more incremental/evolutionary approach will work better, that's definitely worth discussion.

@mkArtakMSFT What do you think? Would you be prepared to put that much investment in this area in .NET 9? You could think of it as being similar to the big project to align the WebAssembly SDKs - we found time for that in .NET 8 so hopefully we have capacity for a similarly large investment in long term code quality in .NET 9.

Reply:
@SteveSandersonMS I definitely agree that a thorough rework of the JS/TS code is required, and I think this is something we should prioritize early in .NET 9, especially if we intend to do some of the other tracked `blazor.*.js` improvements to support things like microfrontend scenarios, etc.

Regarding circuit management, I've decided to update the implementation now to significantly clean up `Boot.Server.Common.ts`. A lot has been moved to `CircuitManager.ts`, which now logically represents the state and lifetime of a single circuit (whose connection can be closed and reinitialized an arbitrary number of times). While it's a fairly large change, it should be easier to reason about and fix bugs going forward.